src/main/jbake/content/blog/2017/06/9/opentracing-spring-boot.adoc

```diff
@@ -0,0 +1,149 @@
+
+== Create a web application
+First let's write a simple web app, no let's generate it!
+
+ifndef::env-github[]
+image::/img/blog/2017/spring-initializr.png[caption="Figure 1: ", title="Spring boot generator"]
+endif::[]
+ifdef::env-github[]
+image::../../../../../assets/img/blog/2017/spring-initializr.png[caption="Figure 1: ", title="Spring boot generator."]
+endif::[]
+
+Now the application is generated, but it does not contain any web controller. We are going to implement simple
+controller with two methods. One will return hello world string and the other creates HTTP request which calls
+hello endpoint. This demonstrates simple request chaining between services. Do not worry all the code is
+on https://github.com/pavolloffay/opentracing-java-examples[GitHub]. At the bottom of this article you will
+find all necessary links.
+
+.Hello Controller:
+----
+@RestController
+public class HelloController {
+
+    @Autowired
+    private RestTemplate restTemplate;
+
+    @RequestMapping("/hello")
+    public String hello() {
+        return "Hello from Spring Boot!";
+    }
+
+    @RequestMapping("/chaining")
+    public String chaining() {
+        ResponseEntity<String> response = restTemplate.getForEntity("http://localhost:8080/hello", String.class);
+        return "Chaining + " + response.getBody();
+    }
+}
+----
+
+Now the application can serve requests for URLs `http://localhost:8080/hello` and `http://localhost:8080/chaining`.
+The app is still not instrumented, we won't see any data coming to a tracing system.
+
+== Instrumentation
+Instrumentation with OpenTracing integrations is very simple. For Spring Boot there is
+an auto-configuration which instruments all REST controllers and `RestTemplate` beans. Just add following
+dependency to the classpath:
+
+----
+<dependency>
+    <groupId>io.opentracing.contrib</groupId>
+    <artifactId>opentracing-spring-web-autoconfigure</artifactId>
+</dependency>
+----
+
+This dependency requires only one thing and that is tracer bean which will be used to report data to the chosen
+tracing system. If we don't specify this bean auto-configuration will choose `NoopTracer`.
+
+Because we are using OpenTracing instrumentation we are not bound to any specific tracing system. If we now choose for
+example Zipkin it is just a matter of configuration change to switch to a different system for instance Jeager or
+LightStep.
+
+As we mentioned the only tracing configuration needed here is to provide a tracer bean. First we are going to Jaeger and
+then Zipkin tracer.
+
+=== Jaeger
+To create Jaeger tracer is very simple. It just requires sampler configuration and because it is a demo we are going
+to sample all requests. Note that we are not specifying URL to Jaeger server. By default it will assume that it
+runs on the localhost.
+
+----
+@Bean
+public io.opentracing.Tracer jaegerTracer() {
+    return new Configuration("spring-boot", new Configuration.SamplerConfiguration(ProbabilisticSampler.TYPE, 1),
```

Comment:
I have used prob sampler in the video, I will stick to that, anyway it's doing the same thing.